GVFS/GVFS.Virtualization/Projection/FileTypeAndMode.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+namespace GVFS.Virtualization.Projection
+{
+    public struct FileTypeAndMode
+    {
+        public static readonly ushort FileMode755;
+        public static readonly ushort FileMode664;
+        public static readonly ushort FileMode644;
+        public static readonly FileTypeAndMode Regular644File;
+
+        // Bitmasks for extracting file type and mode from the ushort stored in the index
+        private const ushort FileTypeMask = 0xF000;
+        private const ushort FileModeMask = 0x1FF;
+
+        private readonly ushort fileTypeAndMode;
+
+        static FileTypeAndMode()
+        {
+            FileMode755 = Convert.ToUInt16("755", 8);
+            FileMode664 = Convert.ToUInt16("664", 8);
+            FileMode644 = Convert.ToUInt16("644", 8);
+
+            Regular644File = new FileTypeAndMode((ushort)((ushort)FileType.Regular | FileMode644));
+        }
+
+        public FileTypeAndMode(ushort typeAndModeInIndexFormat)
+        {
+            this.fileTypeAndMode = typeAndModeInIndexFormat;
+        }
+
+        public enum FileType : ushort
+        {
+            Invalid = 0,
+
+            Regular = 0x8000,
+            SymLink = 0xA000,
+            GitLink = 0xE000,
+        }
+
+        public FileType Type => (FileType)(this.fileTypeAndMode & FileTypeMask);
```

Comment:
To me, part of the value of creating a new type is to not pay the cost of masking every time we access a property. I would either store it as a ushort and add helper methods for masking (still keeping those details private to the parsing code), or else make this a class with separate fields. I also don't see much value in making this a struct.

Reply:
> I also don't see much value in making this a struct.

The value I saw in using a struct if we avoid having to store reference + data in memory.

What do you see as the downside of using a struct?